File: tests/Pgvector.Tests/NpgsqlTests.cs
using Pgvector.Npgsql;

namespace Pgvector.Tests;

public class NpgsqlTests
{
    [Fact]
    public async Task Main()
    {
        var connString = "Host=localhost;Database=pgvector_dotnet_test";

        var dataSourceBuilder = new NpgsqlDataSourceBuilder(connString);
        dataSourceBuilder.UseVector();
        await using var dataSource = dataSourceBuilder.Build();

        var conn = dataSource.OpenConnection();

        await using (var cmd = new NpgsqlCommand("CREATE EXTENSION IF NOT EXISTS vector", conn))
        {
            await cmd.ExecuteNonQueryAsync();
        }

        conn.ReloadTypes();

        await using (var cmd = new NpgsqlCommand("DROP TABLE IF EXISTS items", conn))
        {
            await cmd.ExecuteNonQueryAsync();
        }

        await using (var cmd = new NpgsqlCommand("CREATE TABLE items (embedding vector(3))", conn))
        {
            await cmd.ExecuteNonQueryAsync();
        }

        await using (var cmd = new NpgsqlCommand("INSERT INTO items (embedding) VALUES ($1), ($2), ($3)", conn))
        {
            var embedding1 = new Vector(new float[] { 1, 1, 1 });
            var embedding2 = new Vector(new float[] { 2, 2, 2 });
            var embedding3 = new Vector(new float[] { 1, 1, 2 });
            cmd.Parameters.AddWithValue(embedding1);
            cmd.Parameters.AddWithValue(embedding2);
            cmd.Parameters.AddWithValue(embedding3);
            await cmd.ExecuteNonQueryAsync();
        }

        await using (var cmd = new NpgsqlCommand("SELECT * FROM items ORDER BY embedding <-> $1 LIMIT 5", conn))
        {
            var embedding = new Vector(new float[] { 1, 1, 1 });
            cmd.Parameters.AddWithValue(embedding);

            await using (var reader = await cmd.ExecuteReaderAsync())
            {
                var embeddings = new List<Vector>();

                while (await reader.ReadAsync())
                {
                    embeddings.Add((Vector)reader.GetValue(0));
                }

                Assert.Equal(new float[] { 1, 1, 1 }, embeddings[0].ToArray());
                Assert.Equal(new float[] { 1, 1, 2 }, embeddings[1].ToArray());
                Assert.Equal(new float[] { 2, 2, 2 }, embeddings[2].ToArray());
            }
        }

        await using (var cmd = new NpgsqlCommand("CREATE INDEX ON items USING ivfflat (embedding vector_l2_ops) WITH (lists = 100)", conn))
        {
            await cmd.ExecuteNonQueryAsync();
        }

        await using (var cmd = new NpgsqlCommand("SELECT $1", conn))
        {
            var embedding = new Vector(new float[16000]);
            cmd.Parameters.AddWithValue(embedding);

            await using (var reader = await cmd.ExecuteReaderAsync())
            {
                await reader.ReadAsync();
                Assert.Equal(embedding.ToArray(), ((Vector)reader.GetValue(0)).ToArray());
            }
        }

        await using (var cmd = new NpgsqlCommand("SELECT $1", conn))
        {
            var embedding = new Vector(new float[65536]);
            cmd.Parameters.AddWithValue(embedding);
            var exception = await Assert.ThrowsAsync<System.OverflowException>(() => cmd.ExecuteReaderAsync());
            Assert.Equal("Value was either too large or too small for a UInt16.", exception.Message);
        }
    }
}

File: tests/Pgvector.VisualBasic.Tests/NpgsqlTests.vb
Imports Microsoft.VisualStudio.TestTools.UnitTesting
Imports Microsoft.VisualStudio.TestTools.UnitTesting.CollectionAssert
Imports Npgsql

Namespace Pgvector.VisualBasic.Tests
    <TestClass>
    Public Class NpgsqlTests
        <TestMethod>
        Sub TestSub()
            Dim connString = "Host=localhost;Database=pgvector_dotnet_test"

            Dim dataSourceBuilder As New NpgsqlDataSourceBuilder(connString)
            dataSourceBuilder.UseVector()
            Dim dataSource = dataSourceBuilder.Build()

            Dim conn = dataSource.OpenConnection()

            Using cmd As New NpgsqlCommand("CREATE EXTENSION IF NOT EXISTS vector", conn)
                cmd.ExecuteNonQuery()
            End Using

            conn.ReloadTypes()

            Using cmd As New NpgsqlCommand("DROP TABLE IF EXISTS vb_items", conn)
                cmd.ExecuteNonQuery()
            End Using

            Using cmd As New NpgsqlCommand("CREATE TABLE vb_items (id serial PRIMARY KEY, embedding vector(3))", conn)
                cmd.ExecuteNonQuery()
            End Using

            Using cmd As New NpgsqlCommand("INSERT INTO vb_items (embedding) VALUES ($1), ($2), ($3)", conn)
                Dim embedding1 As New Vector(New Single() {1, 1, 1})
                Dim embedding2 As New Vector(New Single() {2, 2, 2})
                Dim embedding3 As New Vector(New Single() {1, 1, 2})
                cmd.Parameters.AddWithValue(embedding1)
                cmd.Parameters.AddWithValue(embedding2)
                cmd.Parameters.AddWithValue(embedding3)
                cmd.ExecuteNonQuery()
            End Using

            Using cmd As New NpgsqlCommand("SELECT * FROM vb_items ORDER BY embedding <-> $1 LIMIT 5", conn)
                Dim embedding As New Vector(New Single() {1, 1, 1})
                cmd.Parameters.AddWithValue(embedding)

                Using reader As NpgsqlDataReader = cmd.ExecuteReader()
                    Dim ids As New List(Of Integer)
                    Dim embeddings As New List(Of Vector)

                    While reader.Read()
                        ids.Add(reader.GetValue(0))
                        embeddings.Add(reader.GetValue(1))
                    End While

                    CollectionAssert.AreEqual(new Integer() {1, 3, 2}, ids.ToArray())
                    CollectionAssert.AreEqual(new Single() {1, 1, 1}, embeddings(0).ToArray())
                    CollectionAssert.AreEqual(new Single() {1, 1, 2}, embeddings(1).ToArray())
                    CollectionAssert.AreEqual(new Single() {2, 2, 2}, embeddings(2).ToArray())
                End Using
            End Using

            Using cmd As New NpgsqlCommand("CREATE INDEX ON vb_items USING hnsw (embedding vector_l2_ops)", conn)
                cmd.ExecuteNonQuery()
            End Using
        End Sub
    End Class
End Namespace
